### ISSUE TITLE
ResourceQuota object count ignores deletes

### ISSUE THREAD

**Issue by rbubnis** [kubernetes/kubernetes#65107]
**Is this a BUG REPORT or FEATURE REQUEST?**:
/kind bug

**What happened**:
With resource quota defined in the namespace, scaling a deployment down sometimes does not decrement the Pod count on the resourceQuota.  This is seen on pods in deployment or statefulsets, and on service objects. The count will increment in sync with resources, but can skip count updates when deleting resources. In this state, resource counts in all namespaces are affected. A leader change or restarting the kube-controller-mgr container resolves it. Delete/creating of the deployment or namespace does not resolve. It has been left in this state for a many hours without resolving counts.

**What you expected to happen**:
The `used counts` in the resourceQuota status should reflect the current resources when resources are deleted.

```
$ kubectl get po,svc -a  -n test-rkb2
NAME                     READY     STATUS    RESTARTS   AGE
test2-7cc95d6cb8-htzvh   1/1       Running   0          2h

$ kubectl describe quota -n test-rkb2
Name:       resource-quota
Namespace:  test-rkb2
Resource    Used  Hard
--------    ----  ----
pods        3     3
services    1     2
```
Snippet of kube-controller-manager log (--v=6):
```
I0614 13:38:56.301032       1 resource_quota_monitor.go:338] QuotaMonitor process object: /v1, Resource=pods, namespace test-rkb, name test-5b899f499b-lkllz, uid f7d7018e-6fd7-11e8-8f7f-fa163e7097ec, event type delete
...
I0614 13:39:53.673382       1 reflector.go:428] k8s.io/kubernetes/vendor/k8s.io/client-go/informers/factory.go:86: Watch close - *v1.ResourceQuota total 4 items received
I0614 13:39:53.674464       1 round_trippers.go:436] GET https://172.31.0.1:443/api/v1/resourcequotas?resourceVersion=7852&timeoutSeconds=445&watch=true 200 OK in 0 milliseconds
I0614 13:41:49.096798       1 resource_quota_controller.go:193] Resource quota controller queued all resource quota for full calculation of usage
I0614 13:41:53.707197       1 round_trippers.go:436] GET https://172.31.0.1:443/apis/occloud.oracle.com/v1/resourcequotadefaults?resourceVersion=7246&timeoutSeconds=351&watch=true 200 OK in 0 milliseconds
I0614 13:46:49.097070       1 resource_quota_controller.go:193] Resource quota controller queued all resource quota for full calculation of usage
```
After the delete event, I do not see a log entry like the following if in this state:
```
round_trippers.go:436] PUT https://172.31.0.1:443/api/v1/namespaces/test-rkb/resourcequotas/resource-quota/status 200 OK in 3 milliseconds
resource_quota_controller.go:297] Finished syncing resource quota "test-rkb/resource-quota" (238.46085ms)
```

**How to reproduce it (as minimally and precisely as possible)**:
It seems to only occur with a controller taking leadership for the first time (which is why I kill the static pod's container). It is encountered with a test cycle as follows, typically in just a few tries.

Create "test" deployment with replicas=2 in a namespace with resource quota set (delete in step 1 seems to be the trigger):
1.  kubectl scale deploy/test --replicas=1
1. kubectl describe quota
1. if count != 1,  long after pod deleted then errored state is entered
1.  kubectl scale deploy/test --replicas=2
1.  kill kube-controller-manager container
1. continue with step 1

**Anything else we need to know?**:
It seems to only affect controllers taking leadership for the first time (i.e. newly started or just acquired lease but has been running a while).

**Environment**:
- Kubernetes version: 1.9
- Cloud provider or hardware configuration: occurs on multiple hardware/infrastructures - with single or multiple kube-controllers-managers
- OS: Oracle Linux Server 7.5
- Kernel: 4.1.12-124.15.2.el7uek.x86_64

**Comment by rbubnis** [comment]
a guess...
/sig api-machinery

**Comment by liggitt** [comment]
Recovery of quota after delete is done in the background with a reconciling controller, it is not instantaneous

**Comment by rbubnis** [comment]
I monitored logs during testing. The reconciliation seems to run, but not update (I will upload additional logs from a normal vs. stuck cluster). In one instance, I left it in this state for almost 24 hours.

**Comment by rbubnis** [comment]
Full logs at this or a lower level can be uploaded, but here are partial logs (--v=6) for a delete event where the count did not get updated.

Thanks for reviewing
```
I0615 11:09:45.991525       1 flags.go:52] FLAG: --resource-quota-sync-period="5m0s"
I0615 11:09:45.991608       1 controllermanager.go:108] Version: v1.9.1+2.0.2.el7

I0615 11:09:57.728821       1 controllermanager.go:483] Starting "resourcequota"

I0615 11:24:31.574090       1 controller_utils.go:355] Controller test-rkb/test-5b899f499b received delete for pod test-rkb/test-5b899f499b-9bcsm

I0615 11:24:46.429698       1 resource_quota_monitor.go:338] QuotaMonitor process object: /v1, Resource=pods, namespace test-rkb, name test-5b899f499b-9bcsm, uid a114f9ce-708e-11e8-af4f-fa163e8c62c9, event type delete

I0615 11:24:58.959419       1 resource_quota_controller.go:193] Resource quota controller queued all resource quota for full calculation of usage
I0615 11:24:59.222827       1 graph_builder.go:601] GraphBuilder process object: v1/Endpoints, namespace kube-system, name kube-scheduler, uid f1acbb62-7009-11e8-af4f-fa163e8c62c9, event type update
I0615 11:24:59.464552       1 reflector.go:428] k8s.io/kubernetes/vendor/k8s.io/client-go/informers/factory.go:86: Watch close - *v1.Role total 0 items received
I0615 11:24:59.465767       1 round_trippers.go:436] GET https://172.31.0.1:443/apis/rbac.authorization.k8s.io/v1/roles?resourceVersion=992&timeoutSeconds=510&watch=true 200 OK in 1 milliseconds
I0615 11:24:59.524538       1 reflector.go:428] k8s.io/kubernetes/pkg/controller/garbagecollector/graph_builder.go:124: Watch close - <nil> total 0 items received
I0615 11:24:59.527415       1 round_trippers.go:436] GET https://172.31.0.1:443/apis/occloud.oracle.com/v1/externalloggers?resourceVersion=1244&timeoutSeconds=317&watch=true 200 OK in 2 milliseconds

I0615 11:30:48.515770       1 round_trippers.go:436] GET https://172.31.0.1:443/apis/occloud.oracle.com/v1/resourcequotadefaults?resourceVersion=87743&timeoutSeconds=374&watch=true 200 OK in 1 milliseconds
```
and after 30 minutes:
```
$ kubectl describe quota -n test-rkb
Name:       resource-quota
Namespace:  test-rkb
Resource    Used  Hard
--------    ----  ----
pods        2     3
services    0     2

$ kubectl logs kube-controller-manager-ost-bubnis-master6vtl | grep -i quota | [manual filter]
I0615 11:39:58.961195       1 resource_quota_controller.go:193] Resource quota controller queued all resource quota for full calculation of usage
I0615 11:41:57.507991       1 reflector.go:428] k8s.io/kubernetes/vendor/k8s.io/client-go/informers/factory.go:86: Watch close - *v1.ResourceQuota total 0 items received
I0615 11:41:57.509256       1 round_trippers.go:436] GET https://172.31.0.1:443/api/v1/resourcequotas?resourceVersion=87764&timeoutSeconds=422&watch=true 200 OK in 1 milliseconds
I0615 11:44:58.961423       1 resource_quota_controller.go:193] Resource quota controller queued all resource quota for full calculation of usage
I0615 11:45:14.520140       1 round_trippers.go:436] GET https://172.31.0.1:443/apis/occloud.oracle.com/v1/resourcequotadefaults?resourceVersion=87743&timeoutSeconds=319&watch=true 200 OK in 2 milliseconds
I0615 11:48:59.509655       1 reflector.go:428] k8s.io/kubernetes/vendor/k8s.io/client-go/informers/factory.go:86: Watch close - *v1.ResourceQuota total 0 items received
I0615 11:48:59.510517       1 round_trippers.go:436] GET https://172.31.0.1:443/api/v1/resourcequotas?resourceVersion=87764&timeoutSeconds=552&watch=true 200 OK in 0 milliseconds
I0615 11:49:58.961626       1 resource_quota_controller.go:193] Resource quota controller queued all resource quota for full calculation of usage
I0615 11:50:33.523206       1 round_trippers.go:436] GET https://172.31.0.1:443/apis/occloud.oracle.com/v1/resourcequotadefaults?resourceVersion=87743&timeoutSeconds=404&watch=true 200 OK in 2 milliseconds
I0615 11:54:58.961857       1 resource_quota_controller.go:193] Resource quota controller queued all resource quota for full calculation of usage
```

After restarting kube-controller-manager, those logs look like:
```
I0615 11:58:29.332839       1 reflector.go:240] Listing and watching *v1.ResourceQuota from k8s.io/kubernetes/vendor/k8s.io/client-go/informers/factory.go:86
I0615 11:58:29.355225       1 round_trippers.go:436] GET https://172.31.0.1:443/api/v1/resourcequotas?limit=500&resourceVersion=0 200 OK in 22 milliseconds

I0615 11:58:29.361464       1 graph_builder.go:601] GraphBuilder process object: v1/ResourceQuota, namespace test-rkb, name resource-quota, uid 9e51da88-708e-11e8-af4f-fa163e8c62c9, event type add

I0615 11:58:29.767695       1 resource_quota_controller.go:297] Finished syncing resource quota "test/resource-quota" (430.889µs)
I0615 11:58:29.768314       1 resource_quota_controller.go:193] Resource quota controller queued all resource quota for full calculation of usage

I0615 11:58:29.775256       1 round_trippers.go:436] PUT https://172.31.0.1:443/api/v1/namespaces/test-rkb/resourcequotas/resource-quota/status 200 OK in 7 milliseconds
I0615 11:58:29.775312       1 resource_quota_controller.go:297] Finished syncing resource quota "test-rkb/resource-quota" (7.886483ms)

```

**Comment by liggitt** [comment]
it looks like you could be hitting https://github.com/kubernetes/kubernetes/pull/58107, fixed in 1.9.2

https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG-1.9.md#changelog-since-v191
> Fixes a possible deadlock preventing quota from being recalculated (#58107, @ironcladlou)

**Comment by liggitt** [comment]
looking at your log output, that appears to be the deadlock it was hitting. closing this as resolved in #58107 in 1.9.2+

/close
